client/engine/engine.go

```diff
@@ -358,92 +360,89 @@ func (e *Engine) handleChainEvent(chainEvent chainservice.Event) (EngineEvent, e
-			e.store.DestroyConsensusChannel(request.ChannelId)
-			return e.attemptProgress(&ddfo)
+	switch request := or.(type) {
 
-		default:
-			return EngineEvent{}, fmt.Errorf("handleAPIEvent: Unknown objective type %T", request)
+	case virtualfund.ObjectiveRequest:
+		e.metrics.RecordObjectiveStarted(request.Id(*e.store.GetAddress()))
+		vfo, err := virtualfund.NewObjective(request, true, *e.store.GetAddress(), e.store.GetConsensusChannel)
+		if err != nil {
+			return EngineEvent{}, fmt.Errorf("handleAPIEvent: Could not create objective for %+v: %w", request, err)
 		}
 
-	}
-
-	// TODO: Should this live in the payment manager?
-	if cId := apiEvent.PaymentToMake.ChannelId; cId != (types.Destination{}) {
-		voucher, err := e.vm.Pay(
-			cId,
-			apiEvent.PaymentToMake.Amount,
-			*e.store.GetChannelSecretKey())
+		err = e.registerPaymentChannel(vfo)
 		if err != nil {
-			return EngineEvent{}, fmt.Errorf("handleAPIEvent: Error making payment: %w", err)
+			return EngineEvent{}, fmt.Errorf("could not register channel with payment/receipt manager: %w", err)
 		}
-		c, ok := e.store.GetChannelById(cId)
-		if !ok {
-			return EngineEvent{}, fmt.Errorf("handleAPIEvent: Could not get channel from the store %s", cId)
+		return e.attemptProgress(&vfo)
+
+	case virtualdefund.ObjectiveRequest:
+		e.metrics.RecordObjectiveStarted(request.Id(*e.store.GetAddress()))
+		vdfo, err := virtualdefund.NewObjective(request, true, *e.store.GetAddress(), e.store.GetChannelById, e.store.GetConsensusChannel)
+		if err != nil {
+			return EngineEvent{}, fmt.Errorf("handleAPIEvent: Could not create objective for %+v: %w", request, err)
 		}
-		payer, payee := payments.GetPayer(c.Participants), payments.GetPayee(c.Participants)
-		if payer != *e.store.GetAddress() {
-			return EngineEvent{}, fmt.Errorf("handleAPIEvent: Not the sender in channel %s", cId)
+		return e.attemptProgress(&vdfo)
+
+	case directfund.ObjectiveRequest:
+		e.metrics.RecordObjectiveStarted(request.Id(*e.store.GetAddress()))
+		dfo, err := directfund.NewObjective(request, true, *e.store.GetAddress(), e.store.GetChannelsByParticipant, e.store.GetConsensusChannel)
+		if err != nil {
+			return EngineEvent{}, fmt.Errorf("handleAPIEvent: Could not create objective for %+v: %w", request, err)
 		}
-		se := protocols.SideEffects{MessagesToSend: protocols.CreateVoucherMessage(voucher, payee)}
-		err = e.executeSideEffects(se)
+		return e.attemptProgress(&dfo)
+
+	case directdefund.ObjectiveRequest:
+		e.metrics.RecordObjectiveStarted(request.Id(*e.store.GetAddress()))
+		ddfo, err := directdefund.NewObjective(request, true, e.store.GetConsensusChannelById)
 		if err != nil {
-			return EngineEvent{}, fmt.Errorf("handleAPIEvent: Error sending payment voucher: %w", err)
+			return EngineEvent{FailedObjectives: []protocols.ObjectiveId{request.Id(*e.store.GetAddress())}}, fmt.Errorf("handleAPIEvent: Could not create objective for %+v: %w", request, err)
```

Comment:
nit but we could break this up into multiple lines for visibility

Reply:
I'm going to tackle this in a small follow-up PR.

Reply:
#850